server_test.go

```diff
@@ -249,3 +252,208 @@ func TestServerWithBrokenClient(t *testing.T) {
 		srv.Close()
 	}
 }
+
+func TestChroot(t *testing.T) {
+	tmpFolder := "/var/tmp"
+	if runtime.GOOS == "plan9" {
+		tmpFolder = "/tmp"
+	} else if runtime.GOOS == "windows" {
+		tmpFolder = "C:/Windows/Temp"
+	}
+	rootPath, err := ioutil.TempDir(tmpFolder, "sftp")
+	require.Nil(t, err)
+	defer os.RemoveAll(rootPath)
+
+	client, server := clientServerPair(t, Chroot(rootPath))
+	defer client.Close()
+	defer server.Close()
+
+	t.Run("stat", func(t *testing.T) {
+		// prepare, create file and symlink for stat
+		require.Nil(t, os.MkdirAll(filepath.Join(rootPath, "/stat"), 0700))
+		regular := "/stat/regular"
+		symlink := "/stat/symlink"
+		content := []byte(strings.Repeat("hello sftp", 1024))
+		require.Nil(t, ioutil.WriteFile(filepath.Join(rootPath, regular), content, 0700))
+		require.Nil(t, os.Symlink(filepath.Join(rootPath, regular), filepath.Join(rootPath, symlink)))
+		t.Run("regular-stat", func(t *testing.T) {
+			f, err := client.Stat(regular)
+			require.Nil(t, err)
+			require.NotNil(t, f)
+			assert.EqualValues(t, filepath.Base(regular), f.Name())
+			assert.EqualValues(t, len(content), f.Size())
+			assert.True(t, f.Mode().IsRegular())
+			assert.False(t, f.IsDir())
+		})
+		t.Run("symlink-stat", func(t *testing.T) {
+			f, err := client.Stat(symlink)
+			require.Nil(t, err)
+			require.NotNil(t, f)
+			assert.EqualValues(t, filepath.Base(symlink), f.Name())
+			assert.EqualValues(t, len(content), f.Size())
+			assert.True(t, f.Mode().IsRegular())
+			assert.False(t, f.IsDir())
+		})
+		t.Run("regular-lstat", func(t *testing.T) {
+			f, err := client.Lstat(regular)
+			require.Nil(t, err)
+			require.NotNil(t, f)
+			assert.EqualValues(t, filepath.Base(regular), f.Name())
+			assert.EqualValues(t, len(content), f.Size())
+			assert.True(t, f.Mode().IsRegular())
+			assert.False(t, f.IsDir())
+		})
+		t.Run("symlink-lstat", func(t *testing.T) {
+			f, err := client.Lstat(symlink)
+			require.Nil(t, err)
+			require.NotNil(t, f)
+			assert.EqualValues(t, filepath.Base(symlink), f.Name())
+			assert.Greater(t, int64(128), f.Size()) // may have some meta datas
+			assert.NotZero(t, f.Mode()&fs.ModeSymlink)
+		})
+		t.Run("readlink", func(t *testing.T) {
+			f, err := client.ReadLink(symlink)
+			require.Nil(t, err)
+			assert.Equal(t, regular, f)
+		})
+	})
+	t.Run("dir", func(t *testing.T) {
+		require.Nil(t, os.MkdirAll(filepath.Join(rootPath, "/dir"), 0700))
+		assertDir := func(absPath string, exist bool) {
+			f, err := os.Lstat(absPath)
+			if !exist {
+				require.True(t, os.IsNotExist(err))
+			} else {
+				require.Nil(t, err)
+				assert.True(t, f.IsDir())
+			}
+		}
+		t.Run("mkdir", func(t *testing.T) {
+			relPath := "/dir/mkdir"
+			require.Nil(t, client.Mkdir(relPath))
+			// dir should be created
+			assertDir(filepath.Join(rootPath, relPath), true)
```

Comment:
The `*testing.T` used by `assertDir` is captured by closure to be the one from the `t.Run("dir", …)`, but this `*testing.T` is from the `t.Run("mkdir", …)`. Calling functions on the parent `*testing.T` for a daughter `t.Run` is improper usage.